Apps/JobScheduler/src/Jobs/CommunicationJob.cs

```diff
@@ -0,0 +1,167 @@
+                                Email email = new Email()
+                                {
+                                    From = this.fromEmailAddressHGDonotreply,
+                                    To = profile.Email,
+                                    Subject = comm.Subject,
+                                    Body = comm.Text,
+                                    FormatCode = EmailFormat.HTML,
+                                    Priority = comm.Priority,
+                                };
+                                this.emailQueueService.QueueNewEmail(email, false);
+
+                                // Inser a new CommunicationEmail record into
+                                CommunicationEmail commEmail = new CommunicationEmail()
+                                {
+                                    Email = email,
+                                    UserProfile = profile,
+                                    Communication = comm,
+                                };
+                                this.commEmailDelegate.Add(commEmail, false);
+                            }
+
+                            if (usersToSendCommEmails.Count > 0)
+                            {
+                                // Set the createdOnOrAfterFilter for the next query retrieving the next chunk of user profiles.
+                                createdOnOrAfterFilter = usersToSendCommEmails[usersToSendCommEmails.Count - 1].CreatedDateTime;
+                            }
+
+                            if (comm.CommunicationStatusCode != CommunicationStatus.Processing)
+                            {
+                                comm.CommunicationStatusCode = CommunicationStatus.Processing;
+                                this.communicationDelegate.Update(comm, false);
+                            }
+
+                            this.dbContext.SaveChanges(); // commit after every retryFetchSize (or 250) pairs of Email & CommunicationEmail.
+                        }
+                        while (usersToSendCommEmails.Count == this.retryFetchSize
+                               && lastProcessedProfileHdid != usersToSendCommEmails[usersToSendCommEmails.Count - 1].HdId); // keep looping when the above query returns the max return rows (or user profiles).
+
+                        // Update Communication Status to Processed
+                        comm.CommunicationStatusCode = CommunicationStatus.Processed;
+                        this.communicationDelegate.Update(comm, true);
+                    }
+                    catch (Exception e)
+                    {
+                        // log the exception as a warning but we can continue
+                        this.logger.LogWarning($"Error while creating new emails and communication email records for Email Communication - skipping for now\n{e.ToString()}");
+
+                        if (comm.CommunicationStatusCode != CommunicationStatus.Processed)
```

Comment:
This if is unlikely to be ever true (although possible) as the only statements likely to cause an exception are SaveChanges and Update(comm, true).  It is more likely that SaveChanges fails and if that occurs then the update to the comm to the error state is also unlikely to occur.  Hence the reason to set the flag to processing at the start of the loop.